File: src/PicLens.Core/Domain/PathRules.cs
namespace PicLens.Core.Domain;

public static class PathRules
{
    public static StringComparer PathComparer => StringComparer.OrdinalIgnoreCase;

    public static string PathKey(string path) => Path.GetFullPath(path);

    public static bool PathEquals(string? left, string? right) =>
        left is not null
        && right is not null
        && PathComparer.Equals(PathKey(left), PathKey(right));

    public static bool HasSameDirectoryAndBasenameWithoutExtension(string left, string right) =>
        Path.GetDirectoryName(left) is { } leftDirectory
        && Path.GetDirectoryName(right) is { } rightDirectory
        && PathEquals(leftDirectory, rightDirectory)
        && string.Equals(
            Path.GetFileNameWithoutExtension(left),
            Path.GetFileNameWithoutExtension(right),
            StringComparison.OrdinalIgnoreCase);

    public static bool TargetNameExists(IEnumerable<string> existingPaths, string candidatePath, string sourcePath) =>
        existingPaths.Any(path =>
            !PathEquals(path, sourcePath)
            && HasSameDirectoryAndBasenameWithoutExtension(path, candidatePath));
}

File: src/core/src/path_rules.cpp
#include <piclens/core/path_rules.h>

#include <QDir>
#include <QFileInfo>

#include <algorithm>

namespace piclens::core::path_rules {

Qt::CaseSensitivity pathCaseSensitivity()
{
#ifdef Q_OS_WIN
    return Qt::CaseInsensitive;
#else
    return Qt::CaseSensitive;
#endif
}

QString pathKey(const QString &path)
{
    QString key = QDir::cleanPath(QFileInfo(path).absoluteFilePath());
    if (pathCaseSensitivity() == Qt::CaseInsensitive) {
        key = key.toCaseFolded();
    }
    return key;
}

bool pathEquals(const QString &left, const QString &right)
{
    return !left.isNull()
        && !right.isNull()
        && pathKey(left) == pathKey(right);
}

bool hasLinkOrJunctionComponent(const QString &path)
{
    if (path.trimmed().isEmpty()) {
        return false;
    }

    const QFileInfo file(path);
    if (file.isSymLink() || file.isJunction()) {
        return true;
    }

    QDir directory(file.absolutePath());
    while (true) {
        const QFileInfo component(directory.absolutePath());
        if (component.isSymLink() || component.isJunction()) {
            return true;
        }
        if (directory.isRoot() || !directory.cdUp()) {
            return false;
        }
    }
}

bool hasSameDirectoryAndBasenameWithoutExtension(const QString &left, const QString &right)
{
    const QFileInfo leftInfo(left);
    const QFileInfo rightInfo(right);
    return pathEquals(leftInfo.absolutePath(), rightInfo.absolutePath())
        && QString::compare(
               leftInfo.completeBaseName(),
               rightInfo.completeBaseName(),
               pathCaseSensitivity())
            == 0;
}

bool targetNameExists(
    const QVector<QString> &existingPaths,
    const QString &candidatePath,
    const QString &sourcePath)
{
    return std::any_of(existingPaths.cbegin(), existingPaths.cend(), [&](const QString &path) {
        return !pathEquals(path, sourcePath)
            && hasSameDirectoryAndBasenameWithoutExtension(path, candidatePath);
    });
}

} // namespace piclens::core::path_rules
